source/common/upstream/upstream_impl.h

```diff
@@ -178,7 +183,7 @@ class HostDescriptionImpl : virtual public HostDescription,
   Outlier::DetectorHostMonitorPtr outlier_detector_;
   HealthCheckHostMonitorPtr health_checker_;
   std::atomic<uint32_t> priority_;
-  Network::TransportSocketFactory& socket_factory_;
+  std::reference_wrapper<Network::TransportSocketFactory> socket_factory_;
```

Comment:
This needs a `ABSL_GUARDED_BY` annotation as well? Should this use same mutex as metadata?

Reply:
Annotation added . 
Yeah, after checking the code layout I believe that it is possible that metadata update and creating a new upstream connection (which needs access to transport socket factory) to happen at the same time on different threads. To protect against race conditions, a mutex around socket_factory_ has been added.